### DISCUSSION TITLE
Getting a Websocket (I think) error when trying to "finish" the script.

### DISCUSSION THREAD

**Discussion by fanti08** [deepgram/community#580] [General help] [answered]
### Which Deepgram product are you using?

Deepgram API

### Details

When I say stop I have to wait for a couple of seconds, then I get the error and then it stops normally.
### This is the error I'm getting when I say "stop":

> Error while sending: no close frame received or sent
Traceback (most recent call last):
  File "C:\Python312\Lib\site-packages\websockets\sync\connection.py", line 679, in send_context
    self.send_data()
  File "C:\Python312\Lib\site-packages\websockets\sync\connection.py", line 741, in send_data
    self.socket.sendall(data)
  File "C:\Python312\Lib\ssl.py", line 1212, in sendall
    v = self.send(byte_view[count:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python312\Lib\ssl.py", line 1181, in send
    return self._sslobj.write(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2406)
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\Python312\Lib\site-packages\deepgram\audio\microphone\microphone.py", line 130, in _callback
    self.push_callback(input_data)
  File "C:\Python312\Lib\site-packages\deepgram\clients\live\v1\client.py", line 298, in send
    ret = self._socket.send(data)
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python312\Lib\site-packages\websockets\sync\connection.py", line 293, in send
    with self.send_context():
  File "C:\Python312\Lib\contextlib.py", line 144, in __exit__
    next(self.gen)
  File "C:\Python312\Lib\site-packages\websockets\sync\connection.py", line 724, in send_context
    raise self.protocol.close_exc from original_exc
websockets.exceptions.ConnectionClosedError: no close frame received or sent
ERROR: Problem with transcription.
 {'type': 'Exception', 'description': 'Unknown error _listening', 'message': 'cannot join current thread', 'variant': ''}
Exception in _listening: cannot join current thread
Exception in thread Thread-2 (_listening):
Traceback (most recent call last):
  File "C:\Python312\Lib\threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "C:\Python312\Lib\threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Python312\Lib\site-packages\deepgram\clients\live\v1\client.py", line 146, in _listening
    self._emit(
  File "C:\Python312\Lib\site-packages\deepgram\clients\live\v1\client.py", line 112, in _emit
    handler(self, *args, **kwargs)
  File "c:\Users\Utente\Desktop\SPARK\SPARK_tests_1.py", line 22, in on_message
    stop()
  File "c:\Users\Utente\Desktop\SPARK\SPARK_tests_1.py", line 48, in stop
    dg_connection.finish()
  File "C:\Python312\Lib\site-packages\deepgram\clients\live\v1\client.py", line 344, in finish
    self.listening.join()
  File "C:\Python312\Lib\threading.py", line 1144, in join
    raise RuntimeError("cannot join current thread")
RuntimeError: cannot join current thread
Exception ignored in: <module 'threading' from 'C:\\Python312\\Lib\\threading.py'>
Traceback (most recent call last):
  File "C:\Python312\Lib\threading.py", line 1623, in _shutdown
    lock.acquire()
TypeError: ConnectionClosed.__init__() missing 2 required positional arguments: 'rcvd' and 'sent'

### If possible, please attach your code or paste it into the text box.

```
from dotenv import load_dotenv
import os
from deepgram import DeepgramClient, LiveTranscriptionEvents, LiveOptions, Microphone

load_dotenv()

deepgram = DeepgramClient(api_key = os.getenv("deepgram_API"))

stopWord = "stop"
wakeWord = "spark"

def speechRecognition():
    try:
        dg_connection = deepgram.listen.live.v("1")

        def on_message(self, result, **kwargs):
            sentence = result.channel.alternatives[0].transcript

            if len(sentence) == 0:
                return
            if sentence.split()[0].strip('.,!?:;-').lower() == stopWord:
                stop()

            print(f"USER:\n {sentence}")
            
        def on_error(self, error):
            print(f"ERROR: Problem with transcription.\n {error}")
            
        dg_connection.on(LiveTranscriptionEvents.Transcript, on_message)
        dg_connection.on(LiveTranscriptionEvents.Error, on_error)

        options = LiveOptions(
            model = "nova-2-conversationalai",
            smart_format = True,
            language = "en-US",
            encoding = "linear16",
            channels = 1,
            sample_rate = 16000,
        )

        dg_connection.start(options)

        microphone = Microphone(dg_connection.send)
        microphone.start()

        def stop():
            print("STOPPING...")
            dg_connection.finish()
            microphone.finish()

    except Exception as e:
        print(f"ERROR: Exception with transcription.\n {e}")
        return

if __name__ == "__main__":
    speechRecognition()

### If possible, please attach an example audio file to reproduce the issue.

_No response_

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment, marked as answer]
I would take a look at the example:
https://github.com/deepgram/deepgram-python-sdk/blob/main/examples/streaming/microphone/main.py

Notice the:
```
        # wait until finished
        input("Press Enter to stop recording...\n\n")
```

All of these calls are non-blocking in your code... so `main` just exits.

**Reply by fanti08** [reply]
Hey, thanks for the fast reply.
I don't really want to assign a key to the stop recording event, I want to be able to say "Stop" and the program just stops. That's why I'm doing it that way. Do you know a way to do what I just described without getting the error ?

**Reply by davidvonthenen** [reply]
You need to block on something because you don't want to exit your `speechRecognition()` funciton... how this happens is up to you. Another way to put it, how do you want to "signal" to your application you want to "end/finish" the application? Based on your answer, implement that.

**Reply by fanti08** [reply]
I've been trying for long but I can't implement that. That's why I made this post. I tried with 
```
if sentence.split()[0].strip('.,!?:;-').lower() == stopWord:
                stop()
```
but it didnt work because I'm getting that error. That's why I'm asking for help

**Reply by davidvonthenen** [reply]
So you need to block on something... create a global mutex/event that waits indefinitely in `main` after calling `microphone.start()`. In your code:
```
if sentence.split()[0].strip('.,!?:;-').lower() == stopWord:
                # TODO fire said event
```

**Reply by fanti08** [reply]
Thanks after testing and troubleshooting a lot I fixed another way. Thanks for your help anyways

**Reply by lyonsno** [reply]
@fanti08 How did you fix it?
